src/Shared/Server/MethodOptions.cs

```diff
@@ -0,0 +1,160 @@
+
+        /// <summary>
+        /// Gets the maximum message size in bytes that can be sent from the server.
+        /// </summary>
+        public int? MaxSendMessageSize { get; }
+
+        /// <summary>
+        /// Gets the maximum message size in bytes that can be received by the server.
+        /// </summary>
+        public int? MaxReceiveMessageSize { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether detailed error messages are sent to the peer.
+        /// Detailed error messages include details from exceptions thrown on the server.
+        /// </summary>
+        public bool? EnableDetailedErrors { get; }
+
+        /// <summary>
+        /// Gets the compression algorithm used to compress messages sent from the server.
+        /// The request grpc-accept-encoding header value must contain this algorithm for it to
+        /// be used.
+        /// </summary>
+        public string? ResponseCompressionAlgorithm { get; }
+
+        /// <summary>
+        /// Gets the compression level used to compress messages sent from the server.
+        /// The compression level will be passed to the compression provider.
+        /// </summary>
+        public CompressionLevel? ResponseCompressionLevel { get; }
+
+        // Fast check for whether the service has any interceptors
+        internal bool HasInterceptors { get; }
+
+        private MethodOptions(
+            Dictionary<string, ICompressionProvider> compressionProviders,
+            InterceptorCollection interceptors,
+            int? maxSendMessageSize,
+            int? maxReceiveMessageSize,
+            bool? enableDetailedErrors,
+            string? responseCompressionAlgorithm,
+            CompressionLevel? responseCompressionLevel)
+        {
+            CompressionProviders = compressionProviders;
+            Interceptors = interceptors;
+            HasInterceptors = interceptors.Count > 0;
+            MaxSendMessageSize = maxSendMessageSize;
+            MaxReceiveMessageSize = maxReceiveMessageSize;
+            EnableDetailedErrors = enableDetailedErrors;
+            ResponseCompressionAlgorithm = responseCompressionAlgorithm;
+            ResponseCompressionLevel = responseCompressionLevel;
+
+            if (ResponseCompressionAlgorithm != null)
+            {
+                if (!CompressionProviders.TryGetValue(ResponseCompressionAlgorithm, out var _))
+                {
+                    throw new InvalidOperationException($"The configured response compression algorithm '{ResponseCompressionAlgorithm}' does not have a matching compression provider.");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates method options by merging together the settings the specificed <see cref="GrpcServiceOptions"/> collection.
+        /// The <see cref="GrpcServiceOptions"/> should be ordered with items arranged in ascending order of precedence.
```

Comment:
I find this wording to be a little confusing. For settings like MaxSendMessageSize, the last service option will override previous service options which I suppose makes sense for "ascending order of precedence". However, this doesn't quite make sense for interceptors. The pipeline is built in a way that the first service option's interceptors will execute first before the last service option.

Reply:
Added:

> When interceptors from multiple options are merged together they will be executed in reverse order of precendence.